File: app/src/main/java/com/permissionnanny/data/NannyDB.java
package com.permissionnanny.data;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.util.ArrayMap;
import android.support.v4.util.SimpleArrayMap;
import android.text.TextUtils;
import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Map;
import org.iq80.leveldb.DB;
import org.iq80.leveldb.DBIterator;
import org.iq80.leveldb.Options;
import org.iq80.leveldb.WriteBatch;
import org.iq80.leveldb.impl.Iq80DBFactory;

/**
 *
 */
public class NannyDB {
    private static final Charset UTF_8 = Charset.forName("UTF-8");

    private final File mFile;
    private final DB mDB;
    private final Cryo mCryo;

    public NannyDB(File file, Cryo cryo) throws IOException {
        mFile = file;
        Options opt = new Options();
        opt.createIfMissing(true);
        mDB = Iq80DBFactory.factory.open(file, opt);
        mCryo = cryo;
    }

    public void open() {/* No-op: Already open in constructor. */}

    public void close() throws IOException {
        mDB.close();
    }

    public void destroy() throws IOException {
        mDB.close();
        Iq80DBFactory.factory.destroy(mFile, null);
    }

    public <T> T get(@NonNull String key, @NonNull Class<T> type) {
        byte[] data = mDB.get(key.getBytes(UTF_8));
        return mCryo.deserialize(data, type);
    }

    public <T> T get(@NonNull String key) {
        byte[] data = mDB.get(key.getBytes(UTF_8));
        return mCryo.deserialize(data);
    }

    public void put(@NonNull String key, @NonNull Object val) {
        byte[] bytes = mCryo.serialize(val);
        mDB.put(key.getBytes(UTF_8), bytes);
    }

    public void del(@NonNull String key) {
        mDB.delete(key.getBytes(UTF_8));
    }

    public ArrayList<String> findKeys(@NonNull String startsWith) {
        DBIterator it = mDB.iterator();
        it.seek(startsWith.getBytes(UTF_8));
        ArrayList<String> list = new ArrayList<>();
        while (it.hasNext()) {
            list.add(new String(it.next().getKey()));
        }
        return list;
    }

    public <T> ArrayList<T> findVals(@Nullable String startsWith, @NonNull Class<T> type) {
        DBIterator it = mDB.iterator();
        if (!TextUtils.isEmpty(startsWith)) {
            it.seek(startsWith.getBytes(UTF_8));
        }

        ArrayList<T> list = new ArrayList<>();
        while (it.hasNext()) {
            byte[] value = it.next().getValue();
            T data = mCryo.deserialize(value, type);
            if (data != null) {
                list.add(data);
            }
        }
        return list;
    }

    public <V> ArrayMap<String, V> findVal(@Nullable String startsWith, @NonNull Class<V> type) {
        DBIterator it = mDB.iterator();
        if (!TextUtils.isEmpty(startsWith)) {
            it.seek(startsWith.getBytes());
        }

        ArrayMap<String, V> map = new ArrayMap<>();
        while (it.hasNext()) {
            Map.Entry<byte[], byte[]> next = it.next();
            map.put(new String(next.getKey()), mCryo.deserialize(next.getValue(), type));
        }
        return map;
    }

    public <V> void putAtomic(SimpleArrayMap<String, V> map) {
        WriteBatch op = mDB.createWriteBatch();

        for (int i = 0, len = map.size(); i < len; i++) {
            op.put(map.keyAt(i).getBytes(), mCryo.serialize(map.valueAt(i)));
        }
        mDB.write(op);
    }
}

File: app/src/main/kotlin/com/permissionnanny/data/NannyDB.kt
package com.permissionnanny.data

import android.support.v4.util.ArrayMap
import android.support.v4.util.SimpleArrayMap
import android.text.TextUtils
import org.iq80.leveldb.DB
import org.iq80.leveldb.Options
import org.iq80.leveldb.impl.Iq80DBFactory
import java.io.File
import java.io.IOException
import java.nio.charset.Charset
import java.util.*

/**

 */
class NannyDB @Throws(IOException::class)
constructor(
        private val file: File,
        private val cryo: Cryo) {

    private val mDB: DB

    init {
        val opt = Options()
        opt.createIfMissing(true)
        mDB = Iq80DBFactory.factory.open(file, opt)
    }

    fun open() {/* No-op: Already open in constructor. */
    }

    @Throws(IOException::class)
    fun close() {
        mDB.close()
    }

    @Throws(IOException::class)
    fun destroy() {
        mDB.close()
        Iq80DBFactory.factory.destroy(file, null)
    }

    operator fun <T> get(key: String, type: Class<T>): T? {
        val data = mDB.get(key.toByteArray(UTF_8))
        return cryo.deserialize(data, type)
    }

    operator fun <T> get(key: String): T? {
        val data = mDB.get(key.toByteArray(UTF_8))
        return cryo.deserialize(data)
    }

    fun put(key: String, value: Any) {
        val bytes = cryo.serialize(value)
        mDB.put(key.toByteArray(UTF_8), bytes)
    }

    fun del(key: String) {
        mDB.delete(key.toByteArray(UTF_8))
    }

    fun findKeys(startsWith: String): ArrayList<String> {
        val it = mDB.iterator()
        it.seek(startsWith.toByteArray(UTF_8))
        val list = ArrayList<String>()
        while (it.hasNext()) {
            list.add(String(it.next().key))
        }
        return list
    }

    fun <T> findVals(startsWith: String?, type: Class<T>): ArrayList<T> {
        val it = mDB.iterator()
        if (!TextUtils.isEmpty(startsWith)) {
            it.seek(startsWith!!.toByteArray(UTF_8))
        }

        val list = ArrayList<T>()
        while (it.hasNext()) {
            val value = it.next().value
            val data = cryo.deserialize(value, type)
            if (data != null) {
                list.add(data)
            }
        }
        return list
    }

    fun <V> findVal(startsWith: String?, type: Class<V>): ArrayMap<String, V> {
        val it = mDB.iterator()
        if (!TextUtils.isEmpty(startsWith)) {
            it.seek(startsWith!!.toByteArray())
        }

        val map = ArrayMap<String, V>()
        while (it.hasNext()) {
            val next = it.next()
            map.put(String(next.key), cryo.deserialize(next.value, type))
        }
        return map
    }

    fun <V> putAtomic(map: SimpleArrayMap<String, V>) {
        val op = mDB.createWriteBatch()

        var i = 0
        val len = map.size()
        while (i < len) {
            op.put(map.keyAt(i).toByteArray(), cryo.serialize(map.valueAt(i) as Any))
            i++
        }
        mDB.write(op)
    }

    companion object {
        private val UTF_8 = Charset.forName("UTF-8")
    }
}
